backend/src/keyword/keyword.service.ts

```diff
@@ -32,55 +45,53 @@ export class KeywordService extends BaseService<Keyword> {
     if (foundKeyword) {
       throw new ConflictException('Keyword already exists')
     }
-    const jobData = await this.scrapeQueue.add(JOB_NAME, { keyword: tmpEntity.name })
     const { identifiers } = await this.repo.insert(
       this.mapper.map(
         KeywordResultDto,
         Keyword,
         {
           ...tmpEntity,
-          jobQueueId: jobData.id, // NOTE: possibly be `${}${jobId}`
           createdBy
         }
       )
     )
     return identifiers[0]
   }
 
-  async createMany(entities: KeywordResultDto[], createdBy: User): Promise<any> {
+  async createMany(entities: KeywordResultDto[], createdBy: User): Promise<ObjectLiteral[]> {
     const uniqueKeywords = [...new Set(entities.map((entity) => String(entity.name).trim()))]
     const foundKeywords = await this.repo.find({ name: In(uniqueKeywords), createdBy })
     const foundKeywordsStrings = foundKeywords.map((foundKeyword) => foundKeyword.name)
-    const jobs = uniqueKeywords.reduce((vKs, keyword) => {
-      if (foundKeywordsStrings.includes(keyword)) {
-        return vKs
-      }
-      return [...vKs, this.scrapeQueue.add(JOB_NAME, { keyword })]
-    }, [])
-    if (jobs.length) {
-      const allJobs = await Promise.all(jobs)
-      const result = await this.repo.insert(
-        allJobs.map((job) => this.mapper.map(
+    const validKeywordDTOs = uniqueKeywords.reduce((vKs, keyword) => [
+      ...vKs,
+      ...(foundKeywordsStrings.includes(keyword) ? [] : [
+        this.mapper.map(
           KeywordResultDto,
           Keyword,
           {
-            jobQueueId: job.id,
-            name: job.data.keyword,
+            name: keyword,
             createdBy
           }
-        ))
-      )
+        )
+      ])
+    ], [])
+    if (validKeywordDTOs.length) {
+      const result = await this.repo.insert(validKeywordDTOs)
       return result.identifiers
     }
     return []
   }
 
-  findOne(id: EntityId): Promise<Keyword | undefined> {
-    return this.repo.findOne(id, { relations: ['createdBy', 'modifiedBy'] })
-  }
-
-  findOneByJob(jobQueueId: string): Promise<Keyword | undefined> {
-    return this.repo.findOne({ jobQueueId })
+  async findOne(id: EntityId): Promise<Keyword & { queuedJobId?: string } | undefined> {
+    let latestJob = null
+    const jobKeys = await this.scanRedisKeys('0', `${QUEUE_JOB_PREFIX}:${id}-*`, [])
+    if (jobKeys.length) {
+      const latestJobKey = jobKeys.sort().pop()
+      latestJob = await this.scrapeQueue.getJob(latestJobKey.replace(`${QUEUE_JOB_PREFIX}:`, ''))
+    }
```

Comment:
While the removal of the jobQueueID led to uncouple the keyword at the creation stage, there is now coupling at the query level 🤔 Why is there a need to check in Redis for a job? From my understanding, the job class uses `findOne` to query for the keyword record. So there is already a job for that keyword. I might be missing something here so let me know.

Reply:
It's for a technical information section shown on the UI (see attachment), here I tent to show the scraping job status, which could be "In Progress...", "Succeeded" and "Failed" due to the unstableness of my scraping solution . That would lead users to do a "search again" action, somewhere on the UI, which is not implemented yet, it's just my idea 😅.

![Screen_Shot_2022-04-22_at_11_47_07](https://user-images.githubusercontent.com/16256608/164605298-eb70e94b-fe3c-4822-b04b-b0138b453519.jpg)

In that situation, a keyword may have many scraping jobs, but only the latest one need to be shown.

Reply:
Another case, after uploading a CSV file contains a bunch of keywords, let's say 100 of them (it can't be more, because I added a validation for the amount of keywords in a single uploaded file). There will be 100 scraping jobs running sequentially. If users hungry to know a specific keyword's result stats, we might need to know what scraping job is it and then prioritize it order in the jobs queue, or terminate that job and create another one.

Reply:
> It's for a technical information section shown on the UI (see attachment), here I tent to show the scraping job status, which could be "In Progress...", "Succeeded" or "Failed" due to the unstableness of my scraping solution.

The status could be stored at the keyword level:
- `pending` when the keyword has just been created
- `in progress` when the job starts processing the web scraping
- `completed` when the job completed processing the web scraping successfully
- `failed` when the the job failed to process the web scraping

> If users hungry to know a specific keyword's result stats, we might need to know what scraping job is it and then prioritize it order in the jobs queue, or terminate that job and create another one.

That seems more like an entirely new requirement 😅 Providing up-to-date status might be enough for end-users IMO. In any case, since the keyword ID is stored in the job, a search by keyword ID would work too. I see more benefits and simplicity in decoupling jobs from keywords.

Reply:
yeah, agree. I was overthinking a little 😁. You are right, the status should be stored at the keyword entity, and the jobs should update the latest status to keyword entities, that's better. Do you want me to make changes on this ?

Reply:
As explained in [this message](https://github.com/trathailoi/gsrc-exercise/issues/17#issuecomment-1108051395), the code challenge review is now completed. Therefore, it is not required for you to work on making the change for this code challenge.